File: core/src/main/java/de/qabel/core/repository/sqlite/SqliteTransaction.java
package de.qabel.core.repository.sqlite;

import de.qabel.core.repository.Transaction;
import de.qabel.core.repository.exception.TransactionException;

import java.sql.Connection;
import java.sql.SQLException;

public class SqliteTransaction implements Transaction {
    private final Connection connection;
    private final Boolean autocommitState;

    public SqliteTransaction(Connection connection, boolean autocommitState) {
        this.connection = connection;
        this.autocommitState = autocommitState;
    }

    @Override
    public void commit() throws TransactionException {
        try {
            connection.commit();
        } catch (SQLException e) {
            throw new TransactionException("failed to commit", e);
        } finally {
            resetAutocommit();
        }
    }

    @Override
    public void rollback() throws TransactionException {
        try {
            connection.rollback();
        } catch (SQLException e) {
            throw new TransactionException("failed to rollback transaction", e);
        } finally {
            resetAutocommit();
        }
    }

    private void resetAutocommit() {
        try {
            connection.setAutoCommit(autocommitState);
        } catch (SQLException ignored) {
        }
    }

    @Override
    public void close() throws Exception {
        try {
            commit();
        } catch (TransactionException e) {
            rollback();
            throw e;
        }
    }
}

File: core/src/main/java/de/qabel/core/repository/sqlite/SqliteTransaction.kt
package de.qabel.core.repository.sqlite

import de.qabel.core.repository.Transaction
import de.qabel.core.repository.exception.TransactionException

import java.sql.Connection
import java.sql.SQLException

class SqliteTransaction(private val connection: Connection, private val autocommitState: Boolean?) : Transaction {

    @Throws(TransactionException::class)
    override fun commit() {
        try {
            connection.commit()
        } catch (e: SQLException) {
            throw TransactionException("failed to commit", e)
        } finally {
            resetAutocommit()
        }
    }

    @Throws(TransactionException::class)
    override fun rollback() {
        try {
            connection.rollback()
        } catch (e: SQLException) {
            throw TransactionException("failed to rollback transaction", e)
        } finally {
            resetAutocommit()
        }
    }

    private fun resetAutocommit() {
        try {
            connection.autoCommit = autocommitState!!
        } catch (ignored: SQLException) {
        }

    }

    @Throws(Exception::class)
    override fun close() {
        try {
            commit()
        } catch (e: TransactionException) {
            rollback()
            throw e
        }

    }
}
